include/deal.II/grid/grid_tools.h

```diff
@@ -2142,6 +2162,453 @@ namespace GridTools
+            //    gradient.
+            //
+            // There is not much we can do about 1., but for 2. we can derive
+            // finite difference stencils that work on a simplex and rewrite
+            // the optimization problem to use those instead. Consider the
+            // structdim 2 case and let
+            //
+            // F(c0, c1, c2, c3) = Manifold::get_new_point(vertices, {c0, c1, c2, c3})
+            //
+            // then we seek to minimize the Euclidean distance between F(...)
+            // and trial_point. We can solve for c3 in terms of the other
+            // weights and get, for one coordinate direction
+            //
+            // d/dc0 ((x0 - F(c0, c1, c2, 1 - c0 - c1 - c2)))
+            //      = -2(x0 - F(...)) (d/dc0 F(...) - d/dc3 F(...))
+            //
+            // where we substitute back in for c3 after taking the
+            // derivative. We can compute a stencil for the cross derivative
+            // d/dc0 - d/dc3: this is exactly what cross_stencil approximates
+            // (and gradient_entry computes the sum over the independent
+            // variables). Below, we somewhat arbitrarily pick the last
+            // component as the dependent one.
+            //
+            // Since we can now calculate derivatives of the objective
+            // function we can use gradient descent to minimize it.
+            //
+            // Of course, this is much simpler in the structdim = 1 case (we
+            // could rewrite the projection as a 1D optimization problem), but
+            // to reduce the potential for bugs we use the same code in both
+            // cases.
+            auto weights_are_ok = [](const Tensor<1, GeometryInfo<structdim>::vertices_per_cell> &v)
+                                  -> bool
+            {
+              std::array<double, GeometryInfo<structdim>::vertices_per_cell> copied_weights;
+              for (unsigned int i = 0; i < GeometryInfo<structdim>::vertices_per_cell; ++i)
+                {
+                  copied_weights[i] = v[i];
+                  if (v[i] < 0.0 || v[i] > 1.0)
+                    return false;
+                }
+
+              // check the sum: try to avoid some roundoff errors by summing in order
+              std::sort(copied_weights.begin(), copied_weights.end());
+              const double sum = std::accumulate(copied_weights.begin(), copied_weights.end(), 0.0);
+              if (1e-10 < std::abs(sum - 1.0)) // same tolerance used in manifold.cc
+                return false;
+
+              return true;
+            };
+            const double step_size = std::min(object->diameter()/10.0, 1e-6);
+
+            std::vector<Point<spacedim>> vertices;
+            for (unsigned int vertex_n = 0; vertex_n < GeometryInfo<structdim>::vertices_per_cell;
+                 ++vertex_n)
+              vertices.push_back(object->vertex(vertex_n));
```

Comment:
avoid the repeated memory allocation by writing
```
std::vector<Point<spacedim>> vertices (GeometryInfo<structdim>::vertices_per_cell);
for (unsigned int vertex_n = 0; vertex_n < GeometryInfo<structdim>::vertices_per_cell;
       ++vertex_n)
   vertices[vertex_n] = object->vertex(vertex_n);
```
instead.

Reply:
I would prefer not to fix this right now since #4723 will allow us to use a `std::array` here anyway.

Reply:
OK